src/main.cc

```diff
@@ -135,19 +138,41 @@ sig_checkpoint_f(va_list ap)
 	(void)ap;
 	if (box_checkpoint() != 0)
 		diag_log();
+	sig_checkpoint_fiber = NULL;
 	return 0;
 }
 
 static void
 sig_checkpoint(ev_loop * /* loop */, struct ev_signal * /* w */,
 	     int /* revents */)
 {
-	struct fiber *f = fiber_new_system("checkpoint", sig_checkpoint_f);
-	if (f == NULL) {
+	if (sig_checkpoint_fiber != NULL) {
+		say_warn("snapshot is in progress");
+		return;
+	}
+	if (shutdown_started)
+		return;
+	sig_checkpoint_fiber = fiber_new_system("checkpoint", sig_checkpoint_f);
+	if (sig_checkpoint_fiber == NULL) {
 		say_warn("failed to allocate checkpoint fiber");
 		return;
 	}
-	fiber_wakeup(f);
+	fiber_wakeup(sig_checkpoint_fiber);
```

Comment:
This doesn't look like the right place for this check. I think this should be done in `gc_shutdown`.

Reply:
Used gc subsystem to run snapshot triggered by SIGUSR1 as discussed. Now it looks like below:

<details>
<summary>new source patch</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 4985f5a47..316cfe39c 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -5606,13 +5606,17 @@ box_cfg(void)
 int
 box_checkpoint(void)
 {
-	/* Signal arrived before box.cfg{} */
-	if (! is_box_configured)
-		return 0;
-
 	return gc_checkpoint();
 }
 
+void
+box_checkpoint_async(void)
+{
+	if (!is_storage_initialized)
+		return;
+	gc_trigger_checkpoint();
+}
+
 int
 box_backup_start(int checkpoint_idx, box_backup_cb cb, void *cb_arg)
 {
diff --git a/src/box/box.h b/src/box/box.h
index 7bed5efe4..da890a42f 100644
--- a/src/box/box.h
+++ b/src/box/box.h
@@ -228,6 +228,12 @@ box_ro_reason(void);
  */
 int box_checkpoint(void);
 
+/**
+ * Make snapshot asynchronously.
+ */
+void
+box_checkpoint_async(void);
+
 typedef int (*box_backup_cb)(const char *path, void *arg);
 
 /**
diff --git a/src/main.cc b/src/main.cc
index 2bdf67593..1901b3d8a 100644
--- a/src/main.cc
+++ b/src/main.cc
@@ -126,28 +126,14 @@ tarantool_uptime(void)
 	return ev_monotonic_now(loop()) - start_time;
 }
 
-/**
- * Create a checkpoint from signal handler (SIGUSR1)
- */
-static int
-sig_checkpoint_f(va_list ap)
-{
-	(void)ap;
-	if (box_checkpoint() != 0)
-		diag_log();
-	return 0;
-}
-
 static void
 sig_checkpoint(ev_loop * /* loop */, struct ev_signal * /* w */,
 	     int /* revents */)
 {
-	struct fiber *f = fiber_new_system("checkpoint", sig_checkpoint_f);
-	if (f == NULL) {
-		say_warn("failed to allocate checkpoint fiber");
+	say_info("got SIGUSR1 signal bound to checkpoint");
+	if (shutdown_started)
 		return;
-	}
-	fiber_wakeup(f);
+	box_checkpoint_async();
 }
 
 static int

```

</details>